source/as-promise.js

```diff
@@ -27,28 +28,34 @@ module.exports = options => {
 				return;
 			}
 
-			const {statusCode} = response;
 			const limitStatusCode = options.followRedirect ? 299 : 399;
 
 			response.body = data;
 
 			try {
 				for (const hook of options.hooks.afterResponse) {
 					// eslint-disable-next-line no-await-in-loop
-					response = await hook(response);
-
-					if (is.plainObject(response)) {
-						if (emitter.retry(response) === false) {
-							reject(new Error('Retry limit reached.'));
+					response = await hook(response, updatedOptions => {
+						if (emitter.retry(null) === false) {
+							throw new Error('Retry limit reached.');
 						}
-						return;
-					}
+
+						return asPromise(mergeOptions(options, {
+							...updatedOptions,
+							retry: {
+								retries: () => 0
+							},
+							throwHttpErrors: false
```

Comment:
Why this? I think it's surprising behavior that it suddenly doesn't throw on errors.

Reply:
This is used only to get the response object (not an error). The original `asPromise` determines if it should throw or not.

Reply:
here's a test: https://github.com/sindresorhus/got/pull/647/commits/98e51810797c8c93cc0b2fe55f5d0e6d2e7eb687#diff-0920de5c123084fabef750aa0967f8d4R309

also you may set it to `true` and you'll see it fails